One single invocation of intercept can now intercept multiple ports.

thallgren opened the pull request:
It is now possible to intercept multiple ports with one single invocation of `telepresence intercept` by just repeating the `--port` flag.

thallgren committed:
One single invocation of intercept can now intercept multiple ports.

It is now possible to intercept multiple ports with one single
invocation of `telepresence intercept` by just repeating the `--port`
flag.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Add complementary DeleteApp for the NamespacePair Apply app in CI.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Add for multi-port single-intercept tests.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Skip initial tick-delay when pining servers in CI tests.

The `Eventually` function will always wait for its initial tick-delay.
This commit ensures that the PingInterceptedEchoServer pings immediately
and then uses Eventually if that ping fails.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Add check that traffic-manager supports multi-port intercepts.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

thallgren committed:
Ensure that informers are synced when running agent injector test.

Signed-off-by: Thomas Hallgren <thomas@tada.se>

P0lip commented on pkg/client/userd/trafficmgr/intercept.go:
```diff
@@ -333,21 +335,110 @@ func (s *session) ensureNoInterceptConflict(ir *rpc.CreateInterceptRequest) *rpc
 	return nil
 }
 
+func (s *session) compareFinalizedManagerVersion(major, minor, patch uint64) int {
+	mv := s.managerVersion
+	n := mv.Major - major
+	if n == 0 {
+		if n = mv.Minor - minor; n == 0 {
+			n = mv.Patch - patch
+		}
+	}
+	return int(n)
+}
+
```
It might be better to turn it into `requireManagerVersion` and return the boolean value, or use `semver.Version` instead to make it clean.

```suggestion
func (s *session) satisfiesManagerVersion(r semver.Range) bool {
	return r(s.managerVersion)
}
```

and usage

```go
session.satisfiesManagerVersion(semver.MustParseRange(">=2.22.0"))
session.satisfiesManagerVersion(semver.MustParseRange("2.22.0"))
```
etc.


Alternatively, the function could return the severity as well, so major/minor/patch to be more reliable.

Some checks like the one in `ensureWatchers` may yield unexpected results if we have another patch release of traffic-manager 2.20.x (unlikely to happen, but you know what I mean).
As an example, we'll always have the result of `1` if the the traffic-manager version is `2.x.1` while inputs are `2.x.0` and `2.x-1.1`, as an example: `2.21.1` (version), `2.21.0`, `2.20.0`.

thallgren replied:
The problem with semver comparison is that a pre-release or build-release is less than the final release (hence the word `Finalized` in the function). The traffic-manager version isn't final during development.

So let's say I want to check that the traffic-manager is >=2.22.0. I can't just do that with a version range, because `2.22.0-alpha.10` is actually less. The alternative would be to use `>=2.21.n` where `n` is a very high value.

Also, it seems a bit redundant to parse well known numbers from a string. But perhaps that's just me being overly CPU-cycle conservative :-)

thallgren committed:
Let genversion pick the latest version from the CHANGELOG.yml file.

The `TELEPRESENCE_VERSION` defaults to the version generated by the
`genversion` command. Prior to this commit, this command would determine
the version from a call to `git describe --tags --match='v*'`, yielding
a version with the latest known major.minor numbers. This means that
when starting a new minor version, the old minor version would still be
used, unless an arbitrary tag was set to change this behavior.

This commit changes the `genversion` command so that it instead picks
the latest version from the CHANGELOG.md. It will still use the
`git describe --tags --match='v*'` command to get the git generated
pre-release hash.

Signed-off-by: Thomas Hallgren <thomas@tada.se>
